File: 3-leetcode/01-easy/dart/1652-defuse-the-bomb.dart
class Solution {
  List<int> decrypt(List<int> code, int k) {
    List<int> res = List.generate(code.length, (_) => 0);

    if (k == 0) {
      return res;
    }

    int sum = 0;
    int start = k > 0 ? 1 : code.length + k;
    int end = k > 0 ? k : code.length - 1;

    for (int i = start; i <= end; i++) {
      sum += code[i];
    }

    for (int i = 0; i < code.length; i++) {
      res[i] = sum;
      sum -= code[start % code.length];
      sum += code[(end + 1) % code.length];
      start++;
      end++;
    }

    return res;
  }
}

File: 3-leetcode/01-easy/js/1652-defuse-the-bomb.ts
const decrypt = (code: number[], k: number): number[] => {
  const len = code.length;
  const res = Array.from({ length: len }, (v, i) => 0);

  if (k == 0) {
    return res;
  }

  let sum = 0;
  let start = k > 0 ? 1 : len + k;
  let end = k > 0 ? k : len - 1;

  for (let i = start; i <= end; i++) {
    sum += code[i];
  }

  for (let i = 0; i < len; i++) {
    res[i] = sum;
    sum -= code[start % len];
    sum += code[(end + 1) % len];
    start++;
    end++;
  }

  return res;
};
